File: Leetcode/Backtracking/77_Combinations.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  void gen(vector<vector<int>> &ans, vector<int> temp, int start, int n,
           int k) {
    if (temp.size() == k) {
      ans.push_back(temp);
      return;
    }
    if (start > n)
      return;
    for (int i = start; i <= n; i++) {
      temp.push_back(i);
      gen(ans, temp, i + 1, n, k);
      temp.pop_back();
    }
  }
  vector<vector<int>> combine(int n, int k) {
    vector<vector<int>> ans;
    vector<int> temp;
    gen(ans, temp, 1, n, k);
    return ans;
  }
};

File: Leetcode_go/77_Combinations.go
package main

func helper(ans [][]int, temp []int, n int, k int, start int) [][]int {
	if len(temp) == k {
		comb := make([]int, len(temp))
		copy(comb, temp)
		ans = append(ans, comb)
		return ans
	}
	if len(temp) > k {
		return ans
	}
	for i := start; i <= n; i++ {
		temp = append(temp, i)
		ans = helper(ans, temp, n, k, i+1)
		temp = temp[:len(temp)-1]
	}
	return ans

}
func combine(n int, k int) [][]int {
	temp := []int{}
	return helper([][]int{}, temp, n, k, 1)
}
